ape/proposer/instruct_by_score.py

```diff
@@ -0,0 +1,217 @@
+            metric (Optional[str], optional): Metric to use for evaluation. Defaults to None.
+
+        Returns:
+            List[Prompt]: A list of proposed prompts.
+        """
+        # This is not necessary. But I kept it here for now.
+        if not self.data_summary and self.trainset:
+            await self.prepare_dataset_summary()
+
+        # This is the part where we evaluate the base prompt.
+        base_evaluation_score, base_evaluation_result = str(await evaluate(base_prompt, return_outputs=True))
+    
+        proposed_instructions = []
+
+        # Generate N new instructions based on evaluation result.
+        for i in range(N):
+            new_instruction = await self.generate_new_instruction(
+                index=i,
+                base_prompt=base_prompt,
+                evaluation_score=base_evaluation_score,
+                evaluation_result=base_evaluation_result,
+                T=T,
+                response_format=base_prompt.response_format,
+                metric=metric,
+            )
+            proposed_instructions.append(new_instruction)
+
+        return proposed_instructions
+
+    async def generate_new_instruction(
+        self,
+        index: int,
+        base_prompt: Prompt,
+        evaluation_score: str,
+        evaluation_result: List[EvaluationResult],
+        T: float,
+        response_format: Optional[ResponseFormat] = None,
+        metric: Optional[str] = None,
+        retry_count: int = 0,
+    ) -> Prompt:
+        """
+        Generate a new instruction based on the base prompt and its evaluation result.
+
+        Args:
+            base_prompt (Prompt): The base prompt.
+            evaluation_score (str): The score of the base prompt.
+            evaluation_result (List[EvaluationResult]): The result of evaluating the base prompt.
+            T (float): Temperature for generation.
+            response_format (Optional[ResponseFormat], optional): Format for the response.
+            metric (Optional[str], optional): Metric used for evaluation.
+            retry_count (int): 현재 재시도 횟수
+
+        Returns:
+            Prompt: A new proposed prompt or the base prompt if generation fails after 3 attempts.
+        """
+
+        # Select a tip
+        if self.use_tip:
+            selected_tip = list(TIPS.values())[index % len(TIPS)]
+        else:
+            selected_tip = ""
+
+        base_prompt_messages = [
+            json.dumps(message) for message in base_prompt.messages
+        ]
+        base_prompt_messages_str = "\n".join(base_prompt_messages)
+
+        try:
+            # Generate new instruction from Prompt.from_filename("gen-instruction-with-eval")
+            # input : base_prompt, evaluation_result, evaluation_function, tip, response_format
+            new_instruction_text = await self.generate_instructions(
+                base_prompt=base_prompt_messages_str,
+                evaluation_result=str(evaluation_result),
```

Comment:
what value should be sent to LLM as `evaluation_result`?
`List[EvaluationResult]`, or some part of it, or final score(float), or everything, or analysis of intermediate values?
@shamuiscoding